design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,205 @@
+
+As a VM owner I want changes that I applied to the VM object to propagate to the VMI if possible.
+
+As a VM owner, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+> **Note:** Below we are only discussing changes to fields that support live updates.
+                  References to `all` fields only refers to fields that can be live-updated. 
+
+## KubeVirt CR
+
+A new field called `vmRolloutStrategy` will be introduced in the KubeVirt CR. This field will control the behavior of the VM objects. Two possible values will be available:
+
+1. `LiveUpdate` means all changes to the VM object will be eventually applied to the VMI.
+    
+2. `Stage` would mean the current familiar behavior remains - all changes to the VM object will be staged and applied after a restart.
+
+In both cases, a condition called `RestartRequired` will be added to the VM to reflect the need for a restart and the reason why.
+
+The default value of the `rolloutStrategy` field if not specified will be `Stage`.
+
+The `vmRolloutStrategy` field will be usable only if the `VMLiveUpdateFeatures` feature gate is enabled.
+
+
+## VM API
+
+The VM level `liveUpdateFeatures` field will be removed.
+VMs using the `liveUpdateFeatures` will be rejected by the API.
+
+If `RestartRequired` condition is present, any additional changes to the VM spec - to live-updatedable or non-liveupdateable fields - will be staged. Thus the VM has to be restarted.
+
+## Changes to CPU/Memory hotplug workflow
+
+### CPU Hot Plug
+Since the LiveUpdateFeatures API will be deprecated the `spec.template.spec.domain.cpu.maxSockets`field of the VM object will become mutable and stays optional.
+If it is not set by the user then the maxSockets value will continue to be internally computed according the cluster-wide `maxHotPlugRatio`, as it is done today.
+
+Changes to `spec.template.spec.domain.cpu.maxSockets` would require a restart of the VM. This would also set the `RestartRequired` condition.
+
+CPU hot plug is triggered by changing `spec.template.spec.domain.cpu.sockets`. This change will be propagated to the VMI object as long as its value is lower than the value of `MaxSockets` on the last VM startup.
+Increasing the value of the `spec.template.spec.domain.cpu.sockets` beyond the `MaxSockets` will require a restart to take effect.
+
+### Memory Hot Plug
+
+The logic of memory hot plug will be similar to CPU hot plug, with the exception of the fields names.
+`spec.template.spec.domain.memory.maxGuest` will become mutable and will remain optional.
+A default value for the `MaxGuest` memory value will continue to be internally calculated based on the cluster setting of `maxHotplugRatio` 
+
+All other hot plug behavior wil remain the same.
+
+## Drawbacks / Limitations
+
+-   No cluster- and VM-level fine-grained control over rollout settings of specific features.
+-   No VM-specific controll over the rollout strategy.
+
+## API 
```

Comment:
How do you know what is the **actual** maximum resources (CPU, memory) available? Is it reported through the status?

Reply:
Yes. The VMI spec has the `spec.domain.cpu.maxSockets` and `spec.domain.memory.maxGuest`. These fields cannot be updated on a running VMI. These can be set only during creation.

Reply:
@vladikr I think the question is if the current (effective) number of cpu/memory is reported in the status?

Reply:
@fabiand hmm, the original question was about `actual **maximum** resources`.
To your question, the current CPU topology and current memory are already being reported in the status.

Reply:
Too many names, but to be clear: actual == current == effective
The **maximum** can still be expressed based on the information available from the domain or pod, no?

Reply:
For the max Values - I would suggest to also check the physical node - as you can never go beyond the physical resources for CPUs and Memory for a single VM (at least you shouldn't). So having a maximum that is the current physical limit makes sense from my pov.

Reply:
Hotplug is implemented today via live migration. If the target pod won't be scheduled if it requests more resources than any of the nodes can offer.